File: src/main/java/com/jcloisterzone/ai/AiPlayer.java
package com.jcloisterzone.ai;

import com.jcloisterzone.Player;
import com.jcloisterzone.action.ConfirmAction;
import com.jcloisterzone.action.PlayerAction;
import com.jcloisterzone.game.GameSetup;
import com.jcloisterzone.game.state.ActionsState;
import com.jcloisterzone.game.state.GameState;
import com.jcloisterzone.io.message.CommitMessage;
import com.jcloisterzone.io.message.PassMessage;
import com.jcloisterzone.io.message.ReplayableMessage;

import io.vavr.Function1;
import io.vavr.collection.Vector;

public interface AiPlayer extends Function1<GameState, ReplayableMessage> {

    default Vector<ReplayableMessage> getPossibleActions(GameState state) {
        ActionsState as = state.getPlayerActions();

//   	 	System.out.println("DEBUG actions = " + as.getActions());
        
        Vector<ReplayableMessage> messages = as.getActions().flatMap(action -> {
          if (action instanceof ConfirmAction) {
            return Vector.of(new CommitMessage());
          } else {
	        return action.getOptions().map(o -> {
//	       	  System.out.println("DEBUG action = " + action);
//	          System.out.println("DEBUG option = " + o);
	          return Helpers.createMessage(action, o);
	        }).toVector();
          }
        });

        if (as.isPassAllowed()) {// && messages.size() == 0) {
            messages = messages.append(new PassMessage());
        }

        return messages;
    }

    static class Helpers {
        @SuppressWarnings({ "rawtypes", "unchecked" })
        public static ReplayableMessage createMessage(PlayerAction action, Object option) {
//       	 	System.out.println("DEBUG2 action = " + action);
//       	 	System.out.println("DEBUG2 option = " + option);
            return action.select(option);
        }
    }
}

File: src/main/ts/com/jcloisterzone/ai/AiPlayer.ts
import { Vector } from "../../../io/vavr/SeqTypes.js";
import { ConfirmAction } from "../action/ConfirmAction.js";
import type { PlayerAction } from "../action/PlayerAction.js";
import { CommitMessage } from "../io/message/CommitMessage.js";
import { PassMessage } from "../io/message/PassMessage.js";
import type { ReplayableMessage } from "../io/message/ReplayableMessage.js";
import type { GameState } from "../game/state/GameState.js";

/** Base for engine-side AI players: picks a move for the active player.
 *  Port of Java's `AiPlayer extends Function1<GameState, ReplayableMessage>`. */
export abstract class AiPlayer {
  /** Compute the move (or first move of a chain) for the given state. */
  abstract apply(state: GameState): ReplayableMessage;

  /** All legal next messages from the current player actions (each action's options
   *  mapped via `select`, ConfirmAction → COMMIT, plus PASS when allowed). */
  getPossibleActions(state: GameState): Vector<ReplayableMessage> {
    const as = state.getPlayerActions()!;
    let messages = as.getActions().flatMap((action) => {
      if (action instanceof ConfirmAction) {
        return Vector.of(new CommitMessage() as unknown as ReplayableMessage);
      }
      const a = action as PlayerAction<unknown>;
      return a
        .getOptions()
        .toVector()
        .map((o) => a.select(o) as ReplayableMessage);
    }) as Vector<ReplayableMessage>;
    if (as.isPassAllowed()) {
      messages = messages.append(new PassMessage() as unknown as ReplayableMessage) as Vector<ReplayableMessage>;
    }
    return messages;
  }
}
